Add DPT

NielsRogge opened the pull request:
# What does this PR do?

This PR adds [DPT](https://arxiv.org/abs/2103.13413), Dense Prediction Transformers, to the library. It's some very nice work from Intel Labs that applies Transformers for dense prediction tasks such as semantic segmentation and depth estimation.

Feel free to play around with the notebook [here](https://colab.research.google.com/drive/177o79Qm8qsjGDQk5TDSS4tawcFNSZsy8?usp=sharing). 

I've defined 3 models:

* `DPTModel`
* `DPTForDepthEstimation`
* `DPTForSemanticSegmentation`.

DPTModel is the backbone only (ViT in this case). The head models use a neck (DPTNeck) combined with a task-specific head (either `DPTDepthEstimationHead` or `DPTSemanticSegmentationHead`).

Important here:
* a neck is an nn.Module that takes a list of tensors and produces another list of tensors.
* a head takes a list of tensors and return `logits`.

To do:

- [x] make sure heads take a list of tensors as input
- [x] add tests for `DPTFeatureExtractor`
- [x] discuss `out_indices` and `in_index` names
- [x] transfer weights to `Intel` organization

Commit:
First draft

Commit:
More improvements

Commit:
Add fusion blocks

Commit:
Make conversion script work for dpt_large

Commit:
Make conversion script work

Commit:
Improve implementation

Commit:
Improve conversion script

Commit:
Add DPTForSemanticSegmentation

Commit:
Make conversion work for semantic segmentation

Commit:
Add tests

Commit:
Remove print statements

Commit:
First draft

Commit:
Redesign neck

Commit:
Improve tests

Commit:
Improve implementation some more

Commit:
Make neck output list of tensors

Commit:
Improve neck and feature extractor

Commit:
Fix integration tests

Commit:
Make more tests pass

Commit:
Make all tests pass

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_15991). All of your documentation changes will be reflected on that endpoint.

Commit:
Add missing config archive map

sgugger commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,163 @@
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+        out_indices (`List[int]`, *optional*, defaults to [2, 5, 8, 11]):
```
```suggestion
        out_indices (`List[int]`, *optional*, defaults to `[2, 5, 8, 11]`):
```

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.config = config
+
+    def _resize_pos_embed(self, posemb, gs_h, gs_w, start_index=1):
```
`posemb` does not mean anything, even abbreviated. `gs_h`, `gs_w` are very mysterious as well. Please pick better variable names.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        out_channels = config.post_process_channels
+
+        self.projects = nn.ModuleList(
+            [
+                nn.Conv2d(
+                    in_channels=config.hidden_size,
+                    out_channels=out_channel,
+                    kernel_size=1,
+                )
+                for out_channel in out_channels
+            ]
```
```suggestion
            [
                nn.Conv2d(in_channels=config.hidden_size, out_channels=out_channel, kernel_size=1)
                for out_channel in out_channels
            ]
```

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        """
+        Inputs: list of torch.FloatTensor, each of shape (B, L + 1, C).
+        """
+        assert isinstance(inputs, list)
```
No new asserts in the code base.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        assert isinstance(inputs, list)
+        out = []
+
+        for i, x in enumerate(inputs):
+            # reshape to (B, C, H, W)
+            x, cls_token = x[:, 1:], x[:, 0]
+            B, L, C = x.shape
+            size = int(math.sqrt(L))
+            x = x.reshape(B, size, size, C)
+            x = x.permute(0, 3, 1, 2).contiguous()
```
Please pick better variable names.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+    ResidualConvUnit, pre-activate residual unit.
+
+    Args:
+        in_channels (int): number of channels in the input feature map.
+        act_cfg (dict): dictionary to construct and config activation layer.
+        norm_cfg (dict): dictionary to construct and config norm layer.
+        stride (int): stride of the first block. Default: 1
+        dilation (int): dilation rate for convs layers. Default: 1.
+        init_cfg (dict, optional): Initialization config dict. Default: None.
```
If writing a docstring, please follow our usual format.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        return out
+
+
+class DPTPreActResidualConvUnit(nn.Module):
```
This names could be better chosen.

FrancescoSaverioZuppichini replied:
`DPTPreActResidualConvUnit` -> `DPTPreActResidualLayer`

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+            self.batch_norm2 = nn.BatchNorm2d(config.channels)
+
+    def forward(self, inputs):
+        inputs_ = inputs.clone()
```
`original_inputs` seems like a better name.
I also don't see why the clone is necessary since the activation is not inplace.

FrancescoSaverioZuppichini replied:
I think you can go with `hidden_state` and call `input_s` `residual`

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+
+    def forward(self, inputs):
+        inputs_ = inputs.clone()
+        x = self.act1(inputs)
```
Better variable name, please.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        self.res_conv_unit2 = DPTPreActResidualConvUnit(config)
+
+    def forward(self, *inputs):
+        x = inputs[0]
```
Better variable name, please.

sgugger replied:
Same for this one.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+
+        self.res_conv_unit2 = DPTPreActResidualConvUnit(config)
+
+    def forward(self, *inputs):
```
Why accept and ignore more than one input? This is going to lead to problems that are hard to debug. This should only accept one input.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        )
+
+
+class DPTViTPooler(nn.Module):
```
Copied from missing.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+    * FeatureFusionBlocks.
+
+    Args:
+        config (dict): config dict.
```
If writing a docstring, please follow our usual format.

Commit:
Add in_index attribute to make heads accept list of tensors

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        # postprocessing
+        self.reassemble_blocks = DPTReassembleBlocks(config)
+        self.post_process_channels = [
+            channel * math.pow(2, i) if config.expand_channels else channel
```
```suggestion
            channel * 2**i if config.expand_channels else channel
```

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+    """Interpolation module."""
+
+    def __init__(self, scale_factor, mode, align_corners=False):
+        """Init.
+        Args:
+            scale_factor (float): scaling
+            mode (str): interpolation mode
+        """
```
Should go on the class docstring above + follow our usual format.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+            scale_factor (float): scaling
+            mode (str): interpolation mode
+        """
+        super(DPTInterpolate, self).__init__()
```
```suggestion
        super().__init__()
```

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        self.align_corners = align_corners
+
+    def forward(self, x):
+        """Forward pass.
+        Args:
+            x (tensor): input
+        Returns:
+            tensor: interpolated data
+        """
```
If you write a docstring, please follow our usual format.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+            tensor: interpolated data
+        """
+
+        x = self.interpolate(
+            x,
+            scale_factor=self.scale_factor,
+            mode=self.mode,
+            align_corners=self.align_corners,
+        )
```
```suggestion
        x = self.interpolate(x, scale_factor=self.scale_factor, mode=self.mode, align_corners=self.align_corners)
```
Fits in one line.

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1130 @@
+        self.mode = mode
+        self.align_corners = align_corners
+
+    def forward(self, x):
```
Better variable name please.

sgugger commented on src/transformers/models/dpt/test.py:
```diff
@@ -0,0 +1,33 @@
+import torch
```
Should not be included in this PR.

sgugger commented on tests/dpt/test_modeling_dpt.py:
```diff
@@ -0,0 +1,412 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
```suggestion
        config, pixel_values, labels = config_and_inputs
```

sgugger commented on tests/dpt/test_modeling_dpt.py:
```diff
@@ -0,0 +1,412 @@
+    attention_mask and seq_length.
+    """
+
+    all_model_classes = (
+        (
+            DPTModel,
+            DPTForDepthEstimation,
+            DPTForSemanticSegmentation,
+        )
+        if is_torch_available()
+        else ()
+    )
```
Fits in less lines (probably one)

sgugger reviewed (commented):
Thanks for your PR!

There are still a lot of things to clean up in the modeling file. Make sure to follow the usual style guidelines we have in place.

Commit:
Apply suggestions from code review

NielsRogge commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1141 @@
+
+DPT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    # TODO update to organization
+    "nielsr/dpt-large",
```
Note for myself: to be updated.

NielsRogge committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Commit:
Apply some more suggestions

Commit:
Add copied from statements

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+
+
+    Args:
+        hidden_size (`int`, *optional*, defaults to 768):
+            Dimensionality of the encoder layers and the pooler layer.
+        num_hidden_layers (`int`, *optional*, defaults to 12):
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        intermediate_size (`int`, *optional*, defaults to 3072):
+            Dimensionality of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+        hidden_act (`str` or `function`, *optional*, defaults to `"gelu"`):
+            The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+            `"relu"`, `"selu"` and `"gelu_new"` are supported.
+        hidden_dropout_prob (`float`, *optional*, defaults to 0.1):
+            The dropout probabilitiy for all fully connected layers in the embeddings, encoder, and pooler.
+        attention_probs_dropout_prob (`float`, *optional*, defaults to 0.1):
+            The dropout ratio for the attention probabilities.
+        initializer_range (`float`, *optional*, defaults to 0.02):
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-12):
+            The epsilon used by the layer normalization layers.
+        image_size (`int`, *optional*, defaults to 384):
+            The size (resolution) of each image.
+        patch_size (`int`, *optional*, defaults to 16):
+            The size (resolution) of each patch.
+        num_channels (`int`, *optional*, defaults to 3):
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
```
DPT is a framework that works with any (normal vision transformer), what about having a `transformer_config`, similar to what was done in maskformer?

NielsRogge replied:
DPT is really tailored to ViT, for instance it assumes that hidden states have a sequence length of num_patches + 1.

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+        out_indices (`List[int]`, *optional*, defaults to `[2, 5, 8, 11]`):
```
We need a better name :) 

What bout `backbone_hidden_states_to_use`?

sgugger replied:
I think you agreed to use `backbone_out_indices` here?

NielsRogge replied:
Ok, updated this

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+            - "project" passes information to the other tokens by concatenating the readout to all other tokens before
+              projecting the
+            representation to the original feature dimension D using a linear layer followed by a GELU non-linearity.
+        post_process_channels (`List[str]`, *optional*, defaults to [96, 192, 384, 768]):
```
We should not use the word `channels`, right? So

`channels` -> `hidden_size` 

Then, if these are features size of each backbone stage, we should have a better name, e.g. `backbone_hidden_sizes`. I think is better to pass `transformer_config` and accessing the `hidden_sizes` from that

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+            The number of channels before fusion.
+        in_index (`int`, *optional*, defaults to -1):
+            The index of the features to use in the heads.
+        expand_channels (`bool`, *optional*, defaults to `False`):
```
I don't understand that this parameter does

NielsRogge replied:
It will expand the `post_process_channels` by 2*(index) if set to True.

For instance, the first postprocess channels would be 96 * 2^0 = 96, the second one would be 192 * 2^1 = 384, the thrid one 384 * 2^2 = 1536, and so on.

NielsRogge replied:
I've removed it!

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+
+    def __init__(self, image_size=224, patch_size=16, num_channels=3, embed_dim=768):
+        super().__init__()
+        image_size = to_2tuple(image_size)
```
what about removing `to_2tuple` that we are coping over for ages?

NielsRogge replied:
The `to_2tuple` thing is used to support non-square sizes. It's taken from timm

FrancescoSaverioZuppichini replied:
how can it supports non-square sizes if it going to create a tuple with `image_size, image_size`?

NielsRogge replied:
Cause `image_size` can actually also be a tuple of `(height, width)`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+class DPTViTSelfAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        if config.hidden_size % config.num_attention_heads != 0 and not hasattr(config, "embedding_size"):
+            raise ValueError(
+                f"The hidden size {config.hidden_size,} is not a multiple of the number of attention "
+                f"heads {config.num_attention_heads}."
+            )
```
It looks like a check on the config, shouldn't be done in the `Config` `__init__`?

NielsRogge replied:
Yes but then I lose the copied from. If we fix this, it should be done in other modules as well.

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        self.pruned_heads = set()
+
+    def prune_heads(self, heads):
+        if len(heads) == 0:
+            return
```
This is copied from, but nit for early return :)

NielsRogge replied:
Same here, if we fix this, it should be done in other modules as well

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        )
+
+
+class DPTReassembleBlocks(nn.Module):
```
`DPTReassembleLayers`? maybe `DPTReassembleStage`

NielsRogge replied:
Renamed to `DPTReassembleStage`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        16. init_cfg (dict, optional): Initialization config dict. Default: None.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+
+        self.config = config
+        out_channels = config.post_process_channels
+
+        self.projects = nn.ModuleList(
+            [
+                nn.Conv2d(in_channels=config.hidden_size, out_channels=out_channel, kernel_size=1)
+                for out_channel in out_channels
+            ]
+        )
+
+        self.resize_layers = nn.ModuleList(
+            [
+                nn.ConvTranspose2d(
+                    in_channels=out_channels[0], out_channels=out_channels[0], kernel_size=4, stride=4, padding=0
+                ),
+                nn.ConvTranspose2d(
+                    in_channels=out_channels[1], out_channels=out_channels[1], kernel_size=2, stride=2, padding=0
+                ),
+                nn.Identity(),
+                nn.Conv2d(
+                    in_channels=out_channels[3], out_channels=out_channels[3], kernel_size=3, stride=2, padding=1
+                ),
+            ]
+        )
+        if config.readout_type == "project":
+            self.readout_projects = nn.ModuleList()
+            for _ in range(len(self.projects)):
+                self.readout_projects.append(
+                    nn.Sequential(nn.Linear(2 * config.hidden_size, config.hidden_size), ACT2FN[config.hidden_act])
+                )
+
```
This code looks like is doing a lot of stuff. 

- It will break if the backbone returns more than 4 features, so you should create the `self.resize_layers` properly. From the paper, they fixed the target feature size to be the ones returned from the third stage of `ViT`. If we want this code to be general, `DPT` should stage a `target_size` and `later_hidden_states` and compute the rescaling factor based on them. 

A more elegant modeling design (IMHO) is to create a `DPTReassembleLayer`

```python
class DPTReassembleLayer(nn.Sequential):
    def __init__(self, target_size, in_size, in_channels, hidden_size, out_channels):
        super().__init__()
        should_upsample = in_size > target_size
        factor = in_size % target_size
        if should_upsample:
            self.resizer = nn.ConvTranspose2d(
                    in_channels=in_channels, out_channels=hidden_size, kernel_size=factor, stride=factor, padding=0
                )
        elif factor == 0:
            self.resizer = nn.Identity()
        elif not should_upsample:
            # so should downsample
            self.resizer = nn.Conv2d(
                    in_channels=in_channels, out_channels=hidden_size, kernel_size=3, stride=factor, padding=1
                )

        self.project = nn.Conv2d(in_channels=hidden_size, out_channels=out_channels, kernel_size=1)
```

Then you can create `n` layers based on the `backbone_hidden_sizes` and also polish up the forward method

NielsRogge replied:
I've now taken this into account.

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+            x = x.reshape(B, size, size, C)
+            x = x.permute(0, 3, 1, 2).contiguous()
+
+            feature_shape = x.shape
+            if self.config.readout_type == "project":
+                x = x.flatten(2).permute((0, 2, 1))
+                readout = cls_token.unsqueeze(1).expand_as(x)
+                x = self.readout_projects[i](torch.cat((x, readout), -1))
+                x = x.permute(0, 2, 1).reshape(feature_shape)
+            elif self.config.readout_type == "add":
+                x = x.flatten(2) + cls_token.unsqueeze(-1)
+                x = x.reshape(feature_shape)
+            x = self.projects[i](x)
+            x = self.resize_layers[i](x)
+            out.append(x)
```
if it is possible, can you add a comment on the `flatten, permute` showing that is going on? e.g. ` reshaping from b c w h -> b c (wh)`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+    def __init__(self, config):
+        super().__init__()
+
+        self.use_batch_norm = config.use_batch_norm
+        self.act1 = nn.ReLU()
+        self.conv1 = nn.Conv2d(
+            config.channels,
+            config.channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            bias=not self.use_batch_norm,
+        )
+
+        self.act2 = nn.ReLU()
+        self.conv2 = nn.Conv2d(
+            config.channels,
+            config.channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            bias=not self.use_batch_norm,
+        )
+
+        if self.use_batch_norm:
+            self.batch_norm1 = nn.BatchNorm2d(config.channels)
+            self.batch_norm2 = nn.BatchNorm2d(config.channels)
```
Design style, naming multiple layers can get out of and quickly. What about having `self.weight` and stacking everything inside a `nn.Sequential`?

NielsRogge replied:
Ok, what to do with the optional batch norm layers? Something like:

nn.BatchNorm2d(config.channels) if use_batch_norm else nn.Identity()?

FrancescoSaverioZuppichini replied:
That could work, keep in mind I am not sure if we always prefer named layers vs unnamed ones (like stacking them in a sequential layer) so take my design comments with a grain of salt

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+            bias=not self.use_batch_norm,
+        )
+
+        self.act2 = nn.ReLU()
```
probably we want to use the `ACT2FUNC` here?

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+
+    def forward(self, *inputs):
+        x = inputs[0]
+        if len(inputs) == 2:
+            if x.shape != inputs[1].shape:
+                res = nn.functional.interpolate(
+                    inputs[1], size=(x.shape[2], x.shape[3]), mode="bilinear", align_corners=False
+                )
+            else:
+                res = inputs[1]
```
I don't understand what is going on here. There is a lot of control flow, maybe a comment?

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+    def __init__(self, config):
+        super().__init__()
+
+        self.use_batch_norm = config.use_batch_norm
+        self.act1 = nn.ReLU()
+        self.conv1 = nn.Conv2d(
+            config.channels,
+            config.channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            bias=not self.use_batch_norm,
+        )
```
Since this is a `Layer`, I think you can ditch the `config` and just pass the parameters you need so it's way easier to read/understand/reuse in other models. Let me know what do you thinkg

NielsRogge replied:
What do you mean here?

FrancescoSaverioZuppichini replied:
I was asking if we should use the config here or pass everything we need from it as parameters. I am not sure if we have a rule in place for how deep should the config passed

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        self.config = config
+
+        # vit encoder
+        self.embeddings = DPTViTEmbeddings(config)
+        self.encoder = DPTViTEncoder(config)
```
which one is the `vit encoder`? The `DPTVitEncoder` or `DPTViTEmbeddings +DPTViTEncoder `. I think you can encapsulate `DPTViTEmbeddings +DPTViTEncoder ` in `DPTVitModel`

NielsRogge replied:
`DPTModel` corresponds to `ViTModel`.

FrancescoSaverioZuppichini replied:
Okay, why is not called `DPTViTModel` then?

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embeddings.patch_embeddings
```
Not really useful IMHO

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # Prepare head mask if needed
+        # 1.0 in head_mask indicate we keep the head
+        # attention_probs has shape bsz x n_heads x N x N
+        # input head_mask has shape [num_heads] or [num_hidden_layers x num_heads]
+        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
```
very long comment, I think that the comment from line 704 -> 707 may go inside the function as docstring

NielsRogge replied:
Yes but this is again from copied from

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+            return_dict=return_dict,
+        )
+
+        sequence_output = encoder_outputs[0]
```
`sequence_output` is a not so descriptive name imho since we are talking about images

NielsRogge replied:
It is actually also used in ViT, and the name makes sense imo, since ViT operates on a sequence of patches.

NielsRogge replied:
Therefore, marking this as resolved.

FrancescoSaverioZuppichini replied:
Makes sense now actually, thank you for the reply

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+
+class DPTNeck(nn.Module):
+    """
+    DPTNeck. A neck is a module that isused between the backbone and the head. It takes a list of tensors as input and
```
```suggestion
DPTNeck. A neck is a module normally used between the backbone and the head. It takes a list of tensors as input and
```

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        features = [self.convs[i](feature) for i, feature in enumerate(features)]
+
+        # fusion blocks
+        output = []
+        for i in range(len(self.fusion_blocks)):
+            if i == 0:
+                out = self.fusion_blocks[i](features[-1])
+            else:
+                out = self.fusion_blocks[i](out, features[-(i + 1)])
+            output.append(out)
```
This can be simplified. Just start you loop one later and put line `796` before it

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        self.fusion_blocks[0].res_conv_unit1 = None  # not sure why this is done in mmseg
+
+    def forward(self, hidden_states: List[torch.Tensor]) -> List[torch.Tensor]:
+        if len(hidden_states) != len(self.post_process_channels):
+            raise ValueError("The number of hidden states should be equal to the number of post-process channels.")
```
can be checked in the config

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        self.fusion_blocks = nn.ModuleList()
+        for _ in range(len(self.convs)):
+            self.fusion_blocks.append(DPTFeatureFusionBlock(config))
+        self.fusion_blocks[0].res_conv_unit1 = None  # not sure why this is done in mmseg
```
No idea either

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        return output
+
+
+class DPTInterpolate(nn.Module):
```
I don't see the need for this class

NielsRogge replied:
It's because it's used in an `nn.Sequential`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        self.config = config
+
+        features = config.channels
+        self.head = nn.Sequential(
+            nn.Conv2d(features, features // 2, kernel_size=3, stride=1, padding=1),
+            DPTInterpolate(scale_factor=2, mode="bilinear", align_corners=True),
+            nn.Conv2d(features // 2, 32, kernel_size=3, stride=1, padding=1),
+            nn.ReLU(True),
+            nn.Conv2d(32, 1, kernel_size=1, stride=1, padding=0),
+            nn.ReLU(True),
+            nn.Identity(),
+        )
```
why are we using `ReLU` `inplace` here?

NielsRogge replied:
Perhaps because it's in an `nn.Sequential`?

NielsRogge replied:
Maybe to save memory?

FrancescoSaverioZuppichini reviewed (commented):
Thanks for working on it, it looks like a very cool model. There some naming convention that must be applied

- `***Block/s` -> `***Layer/s`
- use `Stage` when we stack multiple `***Layers` together
- when possible use `hidden_size/s` instead of `channels`

Some code can use some refactor to increase readability

Commit:
Remove assert

Commit:
Apply suggestions from code review

Commit:
Apply suggestions from code review

Commit:
Remove DPTInterpolate in favor of nn.Upsample

NielsRogge commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+        posemb = torch.cat([posemb_tok, posemb_grid], dim=1)
+
+        return posemb
+
```
You mean an `nn.Module`? Or just a class?

Commit:
Add comments

NielsRogge commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+    This is the configuration class to store the configuration of a [`DPTModel`]. It is used to instantiate an DPT
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the DPT
+    [google/dpt-large](https://huggingface.co/google/dpt-large) architecture.
```
Note: to be updated.

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1127 @@
+
+        self.projection = nn.Conv2d(num_channels, embed_dim, kernel_size=patch_size, stride=patch_size)
+
+    def forward(self, pixel_values):
```
`pixel_values` -> `hidden_size`
`x` -> `hidden_size` :)

NielsRogge replied:
The embedding layer accepts pixel_values as input. It turns them into `hidden_state`.

NielsRogge replied:
Not sure what you mean here

NielsRogge commented on docs/source/model_doc/dpt.mdx:
```diff
@@ -0,0 +1,56 @@
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
```
Have to fill this in

Commit:
Apply suggestions from code review

NielsRogge commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -479,17 +485,13 @@ def forward(self, inputs):
+            Model configuration class defining the model architecture.
     """
 
     def __init__(self, config):
```
Will not do that as per this tweet: https://twitter.com/ThomasViehmann/status/1452199693165998081

Commit:
Apply suggestions from code review

Commit:
Add proposed design

Commit:
Update design

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -826,7 +826,7 @@ def __init__(self, config):
             ACT2FN["relu"],
```
```suggestion
ACT2FN[config.hidden_act],
```
?

NielsRogge replied:
I can't do this, since `config.hidden_act` is set to "gelu" and used in the backbone

Commit:
Add DPTReassembleLayer

Commit:
Add DPTFeatureFusionStage

NielsRogge commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+            The index of the features to use in the heads.
+        expand_channels (`bool`, *optional*, defaults to `False`):
+            Whether to expand the number of channels of the backbone feature maps.
+        use_batch_norm (`bool`, *optional*, defaults to `False`):
```
Ok, renamed it

Commit:
Apply more suggestions from code review

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
```
single-letter variable name 

maybe add a comment about what is going on :)

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+        mixed_query_layer = self.query(hidden_states)
+
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
```
is that a layer? It seems the result, so `keys`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+
+    def forward(self, hidden_states, input_tensor):
+
+        hidden_states = self.dense(hidden_states)
```
I think `hidden_states` -> `hidden_state` and remove `input_tensor`

NielsRogge replied:
Can't change this, to be done in separate PR for all models

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+
+    def forward(
+        self,
+        hidden_states,
```
`hidden_states` -> `hidden_state`

NielsRogge replied:
Can't change this, to be done in separate PR

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+        return out
+
+
+class DPTReassembleLayer(nn.Module):
+    def __init__(self, config, channels, factor):
+        super().__init__()
+        # projection
+        self.project = nn.Conv2d(in_channels=config.hidden_size, out_channels=channels, kernel_size=1)
+
+        # up/down sampling depending on factor
+        if factor > 1:
+            self.resize = nn.ConvTranspose2d(channels, channels, kernel_size=factor, stride=factor, padding=0)
+        elif factor == 1:
+            self.resize = nn.Identity()
+        elif factor < 1:
+            # so should downsample
+            self.resize = nn.Conv2d(channels, channels, kernel_size=3, stride=int(1 / factor), padding=1)
+
+    def forward(self, hidden_state):
+        hidden_state = self.project(hidden_state)
+        hidden_state = self.resize(hidden_state)
+        return hidden_state
```
Neat!

FrancescoSaverioZuppichini reviewed (commented):
Thanks! A couple of comments about single variables and naming

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1103 @@
+            self.layers.append(DPTFeatureFusionLayer(config))
+        self.layers[0].res_conv_unit1 = None
+
+    def forward(self, hidden_states):
+        output = []
+        out = self.layers[0](hidden_states[-1])
+        for i in range(1, len(self.layers)):
+            out = self.layers[i](out, hidden_states[-(i + 1)])
+            output.append(out)
+
+        return output
```
or maybe `hidden_states_fused`

NielsRogge replied:
I'll go for `fused_hidden_states` ;)

NielsRogge commented:
Thanks for your reviews, addressed most comments. Main thing to update is:
* rename `out_indices`
* rename `in_index`.

Ideally we have a name that is going to be used by all vision models.

Commit:
Apply suggestions from code review

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+    def __init__(self, config, channels, factor):
+        super().__init__()
+        # projection
+        self.project = nn.Conv2d(in_channels=config.hidden_size, out_channels=channels, kernel_size=1)
```
```suggestion
self.projection = nn.Conv2d(in_channels=config.hidden_size, out_channels=channels, kernel_size=1)
```

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+        super().__init__()
+
+        self.use_batch_norm = config.use_batch_norm_in_fusion_residual
+        self.act1 = ACT2FN["relu"]
```
```suggestion
self.activation1 = ACT2FN["relu"]
```

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+
+        self.use_batch_norm = config.use_batch_norm_in_fusion_residual
+        self.act1 = ACT2FN["relu"]
+        self.conv1 = nn.Conv2d(
```
```suggestion
self.convolution1 = nn.Conv2d(
```

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+
+    def forward(self, hidden_state: torch.Tensor) -> torch.Tensor:
+        residual = hidden_state
+        x = self.act1(hidden_state)
```
`x` -> `hidden_state`

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+            self.batch_norm1 = nn.BatchNorm2d(config.fusion_hidden_size)
+            self.batch_norm2 = nn.BatchNorm2d(config.fusion_hidden_size)
+
+    def forward(self, hidden_state: torch.Tensor) -> torch.Tensor:
```
why do we have a type here? :)

NielsRogge replied:
To indicate that `hidden_state` refers to a Tensor and `hidden_states` to a list of Tensors.

FrancescoSaverioZuppichini commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+
+        self.project = nn.Conv2d(config.fusion_hidden_size, config.fusion_hidden_size, kernel_size=1, bias=True)
+
+        self.res_conv_unit1 = DPTPreActResidualLayer(config)
```
```suggestion
self.residual_layer1
```

FrancescoSaverioZuppichini reviewed (approved):
Thanks! Approved, there are still some minor comments about variables names

Commit:
Apply suggestions from code review

sgugger commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,167 @@
+            The number of output channels for each of the four feature maps of the backbone.
+        channels (`int`, *optional*, defaults to 256):
+            The number of channels before fusion.
+        in_index (`int`, *optional*, defaults to -1):
```
I think you agreed to use `head_in_indices` here?

sgugger commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1108 @@
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def compute_loss(self, logits, auxiliary_logits, labels):
+        # upsample logits to the images' original size
+        upsampled_logits = nn.functional.interpolate(
+            logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+        )
+        if auxiliary_logits is not None:
+            upsampled_auxiliary_logits = nn.functional.interpolate(
+                auxiliary_logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+            )
+        # compute weighted loss
+        loss_fct = CrossEntropyLoss(ignore_index=self.config.semantic_loss_ignore_index)
+        main_loss = loss_fct(upsampled_logits, labels)
+        auxiliary_loss = loss_fct(upsampled_auxiliary_logits, labels)
+        loss = main_loss + self.config.auxiliary_loss_weight * auxiliary_loss
+
+        return loss
```
We had the debate on another model recently, and IMO this should all be in the forward pass.

NielsRogge replied:
Imo, it's way cleaner to not include this in the forward pass. The same is done in BEiT. (as this one also uses an auxiliary head).

cc @FrancescoSaverioZuppichini

sgugger replied:
I don't find this cleaner, it makes the reader go up and down in the modeling file for 7 lines of code while introducing inconsistency with every other model in the library.

BEiT should also not have defined this separate `compute_loss` function IMO.

FrancescoSaverioZuppichini replied:
I also find this is cleaner and I'd personally prefer to read a file defined in this way (but this is personally). I don't have the experience to tell if it is negligible to have this kind of inconsistency thou.

patrickvonplaten replied:
+1 for Sylvain from my side. I really think it's important to also keep the "Principal of least astonishment": https://en.wikipedia.org/wiki/Principle_of_least_astonishment into consideration here a bit as I've never seen any other model that has a `compute_loss` function. Changing the design here comes with a big cost of "astonishment" IMO for everybody that is used to how the loss is computed previously / in other models

NielsRogge replied:
Ok I have removed the `compute_loss` method, it's now all included in the forward pass.

sgugger reviewed (commented):
There are still a few comments pending from the first review, highlighted them.

Commit:
Fix rebase

Commit:
Update in_index and out_indices

Commit:
Fix conversion script

Commit:
Fix code quality

Commit:
Add model to toctree and use DepthEstimatorOutput

Commit:
Fix code examples

Commit:
Improve code

Commit:
Fix rebase

Commit:
Fix copied from statements

patrickvonplaten commented on src/transformers/models/dpt/__init__.py:
```diff
@@ -0,0 +1,59 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

patrickvonplaten commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,171 @@
+            The number of input channels.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether to add a bias to the queries, keys and values.
+        backbone_out_indices (`List[int]`, *optional*, defaults to `[2, 5, 8, 11]`):
```
Sorry not too familiar with all vision models yet, but has `backbone` become a common term for vision models? What does it stand for exactly?

patrickvonplaten commented on src/transformers/models/dpt/configuration_dpt.py:
```diff
@@ -0,0 +1,171 @@
+        attention_probs_dropout_prob=0.0,
+        initializer_range=0.02,
+        layer_norm_eps=1e-12,
+        is_encoder_decoder=False,
```
(nit) That's false by default. No need to pass it here IMO
```suggestion
```

patrickvonplaten commented on src/transformers/models/dpt/feature_extraction_dpt.py:
```diff
@@ -0,0 +1,202 @@
+        # transformations (resizing + normalization)
+        if self.do_resize and self.size is not None:
+            for idx, image in enumerate(images):
+                size = self.update_size(image)
```
(just to understand a bit better) -> `update_size` never introduces padding values no? Can images be padded? Just curious because if one pads and normalizes afterwards this might lead to problems which is what occurred in the speech feature extraction a while back

NielsRogge replied:
No, images aren't padded.

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+
+    def __init__(self, image_size=224, patch_size=16, num_channels=3, embed_dim=768):
+        super().__init__()
+        image_size = to_2tuple(image_size)
```
(nit) for me it's not worth abstracting away two lines of code

NielsRogge replied:
Yeah we'll remove the `to_2tuple` thing in a separate PR for all vision models.

cc @FrancescoSaverioZuppichini

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+    def __init__(self, config: DPTConfig) -> None:
+        super().__init__()
+        self.chunk_size_feed_forward = config.chunk_size_feed_forward
+        self.seq_len_dim = 1
```
this is never used no?

patrickvonplaten replied:
```suggestion
```

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+        self.layer = nn.ModuleList([DPTViTLayer(config) for _ in range(config.num_hidden_layers)])
+        self.gradient_checkpointing = False
+
+    # Copied from transformers.models.vit.modeling_vit.ViTEncoder.forward
```
Why not copy the whole encoder - is there anything different?

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+        self.layers = nn.ModuleList()
+        for _ in range(len(config.neck_hidden_sizes)):
+            self.layers.append(DPTFeatureFusionLayer(config))
+        self.layers[0].residual_layer1 = None
```
uff that looks a bit hacky. Can't we pass a flag `add_residual_layer1` to `DPTFeatureFusionLayer` with 
`self.layers.append(DPTFeatureFusionLayer(config, add_residual_layer1 = (i > 0)))`

NielsRogge replied:
Turns out it works too when removing this line.

So I won't include it.

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+        self.residual_layer2 = DPTPreActResidualLayer(config)
+
+    def forward(self, hidden_state, residual=None):
+        if residual is not None:
```
(nit) not a huge fan of this if statement and sometimes having `residual_layer1` None sometimes not. Would prefer to have two layer classes, but don't feel strongly about it

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+        >>> inputs = feature_extractor(images=image, return_tensors="pt")
+
+        >>> outputs = model(**inputs)
+        >>> predicted_depth = outputs.predicted_depth  # shape (batch_size, height, width)
```
maybe make it a doc test? :-)

patrickvonplaten replied:
E.g. adding an expected result

patrickvonplaten commented on src/transformers/models/dpt/modeling_dpt.py:
```diff
@@ -0,0 +1,1121 @@
+            if self.config.num_labels == 1:
+                raise ValueError("The number of labels should be greater than one")
+            else:
+                loss = self.compute_loss(logits, auxiliary_logits, labels)
```
Having followed until here, the reader now has to CTRL+F `compute_loss` which she/he has never done for other models

patrickvonplaten commented on tests/dpt/test_feature_extraction_dpt.py:
```diff
@@ -0,0 +1,191 @@
+    def feat_extract_dict(self):
+        return self.feature_extract_tester.prepare_feat_extract_dict()
+
+    def test_feat_extract_properties(self):
```
(out of curiousity) Why have this test ? Do all image models require those attributes? Should it then maybe go into a general mixin class?

patrickvonplaten commented on tests/dpt/test_feature_extraction_dpt.py:
```diff
@@ -0,0 +1,191 @@
+        self.assertTrue(hasattr(feature_extractor, "do_resize"))
+        self.assertTrue(hasattr(feature_extractor, "size"))
+
+    def test_batch_feature(self):
```
Think decorator skip statements are better ;-)

patrickvonplaten commented on tests/dpt/test_modeling_dpt.py:
```diff
@@ -0,0 +1,399 @@
+@require_torch
+@require_vision
+@slow
+class DPTModelIntegrationTest(unittest.TestCase):
```
Cool !

patrickvonplaten commented on tests/dpt/test_modeling_dpt.py:
```diff
@@ -0,0 +1,399 @@
+            loss.backward()
+
+    def test_training_gradient_checkpointing(self):
+        if not self.model_tester.is_training:
```
```suggestion
        if not self.model_tester.is_training:
```
We can remove this here no? The test always runs no?

patrickvonplaten commented on utils/documentation_tests.txt:
```diff
@@ -49,3 +49,17 @@ src/transformers/models/wav2vec2/modeling_wav2vec2.py
 src/transformers/models/wav2vec2/tokenization_wav2vec2.py
 src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py
 src/transformers/models/wavlm/modeling_wavlm.py
+src/transformers/models/bart/modeling_bart.py
```
This seems wrong? They are already added in master - could you rebase here?

patrickvonplaten reviewed (approved):
Looks good to me in general! 

Just three points I feel somewhat strongly about:

- It'd be nice to add `# Copied from` at the highest level (i.e. the class level) possible IMO and not just the forward method
- Don't like the `compute_loss` function abstraction, forces the reader to jump between functions unnecessarily & is not in line with other models
- Think there is a problem with the doc tests documents file -> all those model are in master already I think so no need to add them. Would be nice to add DPT though

Commit:
Apply suggestions from code review

Commit:
Remove compute_loss method

Commit:
Apply suggestions from code review

Commit:
Fix documentation tests file

Commit:
Remove test.py file

Commit:
Improve doc example

debtriche commented:
stop with all the emails I’m getting mad



Sent from my iPhone

> On Mar 11, 2022, at 8:37 AM, NielsRogge ***@***.***> wrote:
> 
> ﻿
> @NielsRogge commented on this pull request.
> 
> In src/transformers/models/dpt/modeling_dpt.py:
> 
> > +            if output_attentions:
> +                all_self_attentions = all_self_attentions + (layer_outputs[1],)
> +
> +        if output_hidden_states:
> +            all_hidden_states = all_hidden_states + (hidden_states,)
> +
> +        if not return_dict:
> +            return tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)
> +        return BaseModelOutput(
> +            last_hidden_state=hidden_states,
> +            hidden_states=all_hidden_states,
> +            attentions=all_self_attentions,
> +        )
> +
> +
> +class DPTReassembleBlocks(nn.Module):
> Renamed to DPTReassembleStage
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you are subscribed to this thread.
